libsolidity/ast/Types.cpp

```diff
@@ -604,6 +601,12 @@ FixedPointType::FixedPointType(int _totalBits, int _fractionalDigits, FixedPoint
 		"Invalid bit number(s) for fixed type: " + 
 		dev::toString(_totalBits) + "x" + dev::toString(_fractionalDigits)
 	);
+
+	unsigned const bitsAvailable = m_totalBits - (isSigned() ? 1 : 0);
+	solAssert(
+		fitsPrecisionExp(10, m_fractionalDigits, bitsAvailable),
```

Comment:
This should actually become a static helper and be used in the typechecker too, to restrict invalid combinations:
```
static FixedPointType::isValid(unsigned _totalBits, unsigned _fractionalDigits, bool _isSigned)
{
    unsigned const bitsAvailable = _totalBits - (_isSigned ? 1 : 0);
    return fitsPrecisionExp(10, _fractionalDigits, bitsAvailable);
}
```

Reply:
Actually I think it may be better to check, whether 10^m_fractionalDigits <= 2^bitsAvailable directly here. The motivation behind ``fitsPrecisionExp`` was to restrict exponentiation of literals to **roughly** 4096 bits - and we didn't want to do the exponentiation before the check, since calculating it could already have eaten up a lot of memory. However, the calculation is approximative (maybe that should be documented...). That's not a problem for literals (the idea was to cut off at some random insanely high value that's *around* 2^4096), but here (1) precision does matter and (2) ``m_fractionalDigits`` and ``m_totalBIts`` (resp. ``bitsAvailable``) is limited to reasonably low values anyways, so doing the precise calculation doesn't hurt.

Reply:
I would even say that we could just pre-calculate the max digits supported for every bit width between 8 and 256 in 8 bit increments.

Reply:
Well, we could do that... but it may be more difficult to see what's going on and whether the values are correct that way. I would say something like ``boost::multiprecision::exp(bigint(10), m_fractionalDigits) <= boost::multiprecision::exp(bigint(2), bitsAvailable)`` doesn't hurt too much.

Actually, in general I think it's rather weird that fixed byte types use a fixed number of decimal digits for the fractional part instead of a fixed fractional bit resolution, but judging from the related issues it seems you had discussed and decided that before my time :).

Reply:
It was fractional bits before, but it seems to be more natural to know number of decimal digits in calculations as a user.

It is more painful in some other parts from a tooling perspective.

Reply:
Ah I remember the limitation was that `exp` was not available in the boost we're using. Any suggestion?

Reply:
Ah, ok. Actually, for that I would have used a table in the docs that tells you the number of decimal digits a given amount of fractional bits gets you :). But if users want it this way, then we should of course do it this way (even though I imagine that not only tooling will be more painful, but conversions and arithmetic using fixed point numbers might also cost more gas this way in the end).

Reply:
Ehm, sorry, I think it should be ``pow``, not ``exp``, but that should be in boost.

Reply:
So either ``boost::multiprecision::pow(bigint(10), m_fractionalDigits) <= boost::multiprecision::pow(bigint(2), bitsAvailable)``
or ``boost::multiprecision::pow(bigint(10), m_fractionalDigits) <= boost::multiprecision::pow2(bitsAvailable)``
or even ``boost::multiprecision::pow(bigint(10), m_fractionalDigits) <= (bigint(1) << bitsAvailable)`` should all be fine.